b9lab-content/3-main-concepts/10-multistore-keepers.md

```diff
@@ -108,27 +108,357 @@ The AnteHandler is theoretically optional, but still a very important component
+
+* Get the next game id:
+    ```go
+    nextGame, found := k.GetNextGame(ctx)
+    if !found {
+        // Panic because this should never happen.
+        panic("NextGame not found")
+    }
+    newIndex := strconv.FormatUint(nextGame.IdValue, 10)
+    ```
+* Extract and sanitize the message elements:
+    ```go
+    creator, err := sdk.AccAddressFromBech32(msg.Creator)
+    if err != nil {
+        return nil, errors.New("Creator address invalid")
+    }
+    red, err := sdk.AccAddressFromBech32(msg.Red)
+    if err != nil {
+        // Standard error because users can make mistakes.
+        return nil, errors.New("Red address invalid")
+    }
+    black, err := sdk.AccAddressFromBech32(msg.Black)
+    if err != nil {
+        return nil, errors.New("Black address invalid")
+    }
+    ```
+* Create the stored game object:
+    ```go
+    storedGame := types.StoredGame{
+        Creator: msg.Creator,
+        Index:   newIndex,
+        Game:    rules.New().String(),
+        Red:     msg.Red,
+        Black:   msg.Black,
+        Wager:   msg.Wager,
+    }
+    ```
+* Save it in storage:
+    ```go
+    k.SetStoredGame(ctx, storedGame)
+    ```
+* Prepare for the next created game:
+    ```go
+    nextGame.IdValue++
+    k.SetNextGame(ctx, nextGame)
+    ```
+* Return the game id for reference:
+    ```go
+    return &types.MsgCreateGameResponse{
+        IdValue: newIndex,
+    }, nil
+    ```
+
+As for `MsgPlayMoveResponse` and `MsgRejectGame`, you would do the same and is left as an exercise.
+
+## More on Game Theory
+
+When you introduced the [messages](./07-messages), there remained the question of what to do when a player does not play ball.
+
+Time to introduce the game deadline:
+
+```go
+const (
+    MaxTurnDurationInSeconds = time.Duration(24 * 3_600 * 1000_000_000) // 1 day
+    DeadlineLayout           = "2006-01-02 15:04:05.999999999 +0000 UTC"
+)
+type StoredGame struct {
+    ...
+    Deadline string
+}
+```
+Set its initial value on creation:
+
+```go
+storedGame := types.StoredGame{
+    ...
+    Deadline: ctx.BlockTime().Add(types.MaxTurnDurationInSeconds).UTC().Format(types.DeadlineLayout),
+}
+```
+Update its value after a move:
+
+```go
+storedGame.Deadline = ctx.BlockTime().Add(types.MaxTurnDurationInSeconds).UTC().Format(types.DeadlineLayout)
+```
+Extract and verify its value when necessary:
+
+```go
+deadline, err = time.Parse(DeadlineLayout, storedGame.Deadline)
+if err != nil {
+    panic(err)
+}
+if deadline.Before(ctx.BlockTime()) {
+    // TODO
+}
+```
+
+## How To Expire Games
+
+Ok, but when do you verify that the game has expired? An interesting feature of an ABCI application is that you can do some action at the end of each block. Should you load all games and filter for those that have expired? No. That would be extremely expensive. Better keep a FIFO where fresh games are pushed back to the tail and therefore the head contains the next games to expire.
```

Comment:
```suggestion
Ok, but when do you verify that the game has expired? An interesting feature of an ABCI application is that you can have it perform some actions at the end of each block. Should you load all games and filter for those that have expired? No. That would be extremely expensive. Better keep a FIFO where fresh games are pushed back to the tail so that the head contains the next games to expire.
```